os_repository.go

```diff
@@ -0,0 +1,202 @@
+// Copyright © 2024 Dell Inc. or its subsidiaries. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//      http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package goscaleio
+
+import (
+	"bytes"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"time"
+
+	types "github.com/dell/goscaleio/types/v1"
+)
+
+// GetAllOSRepositories Gets all OS Repositories
+func (gc *GatewayClient) GetAllOSRepositories() ([]types.OSRepository, error) {
+	defer TimeSpent("GetAllOSRepositories", time.Now())
+
+	path := "/api/v1/OSRepository"
+	req, httpError := http.NewRequest(http.MethodGet, gc.host+path, nil)
+	if httpError != nil {
+		return nil, httpError
+	}
+
+	if gc.version == "4.0" {
+		req.Header.Set("Authorization", "Bearer "+gc.token)
+
+		err := setCookie(req.Header, gc.host)
+		if err != nil {
+			return nil, fmt.Errorf("Error While Handling Cookie: %s", err)
+		}
+	} else {
+		req.Header.Set("Authorization", "Basic "+base64.StdEncoding.EncodeToString([]byte(gc.username+":"+gc.password)))
+	}
+
+	req.Header.Set("Content-Type", "application/json")
+
+	client := gc.http
+	httpResp, httpRespError := client.Do(req)
+	if httpRespError != nil {
+		return nil, httpRespError
+	}
+
+	if httpResp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("OS Repositories not found")
+	}
+	var osRepositories []types.OSRepository
+	responseString, _ := extractString(httpResp)
+	err := json.Unmarshal([]byte(responseString), &osRepositories)
+	if err != nil {
+		return nil, fmt.Errorf("Error parsing response data for OS repositories: %s", err)
+	}
+	return osRepositories, nil
+}
+
+// GetOSRepositoryByID Gets OS Repository by ID
+func (gc *GatewayClient) GetOSRepositoryByID(id string) (*types.OSRepository, error) {
+	defer TimeSpent("GetOSRepositoryByID", time.Now())
+
+	path := fmt.Sprintf("/api/v1/OSRepository/%v", id)
+	req, httpError := http.NewRequest(http.MethodGet, gc.host+path, nil)
+	if httpError != nil {
+		return nil, httpError
+	}
+
+	if gc.version == "4.0" {
```

Comment:
One basic question, Is Bearer token concept is only for v4.0 or version >= 4.0.
Can we consider this so code will be future proof?

Reply:
Yes I believe so. It's supported for versions above 4.0 and used in many other Resources like deploy, service, template etc.

Reply:
Could you please confirm this with the backend storage/array team?

Regarding versioning, please consider both minor and patch versions. Even if this is intended for v4.0, it should also apply to all 4.x.x versions. Additionally, would it be possible to include some constants for this as well?

Thanks

Reply:
@adarsh-dell I have updated code to remove specifying the exact  version before calling the APIs. It would call the API and fail/pass based on what's supported and what's not.